dashboard/src/components/AppView/CustomAppView/CustomAppView.tsx

```diff
@@ -0,0 +1,28 @@
+import { InstalledPackageDetail } from "gen/kubeappsapis/core/packages/v1alpha1/packages";
+import { useMemo } from "react";
+import { useSelector } from "react-redux";
+import { CustomComponent } from "RemoteComponent";
+import { IStoreState } from "shared/types";
+import { IAppViewResourceRefs } from "../AppView";
+
+export interface ICustomAppViewProps {
+  resourceRefs: IAppViewResourceRefs;
+  app: InstalledPackageDetail;
+}
+
+function CustomAppView({ resourceRefs, app }: ICustomAppViewProps) {
+  const {
+    config: { remoteComponentsUrl },
+  } = useSelector((state: IStoreState) => state);
+
+  const url = remoteComponentsUrl
+    ? remoteComponentsUrl
+    : `${window.location.origin}/custom_components.js`;
```

Comment:
The naming here implies that the remote URL (and the default local url) defines custom component*s* (plural), so the bundle defines multiple custom components to be used by different parts of the code (app views, forms for deployment etc.), but I can't see anything being passed to the CustomComponent below to identify a specific custom component from the bundle? What am I missing?

Reply:
Yeah I'm glad you pointed this out. I definitely need to add some sort of identifier to make it easier to determine what to render.

Reply:
It's still not clear to me how the custom component below, which is given only the url to the bundle of remote component, will be able to identify and use the specific remote component from the bundle. Or I've completely misunderstood what the intent was. That's why I'm keen to see an example bundle with multiple components, to understand how you're planning to use multiple remote components for different packages/charts, before we land this.

Looking at the README at https://github.com/Paciolan/remote-component makes it look like it's a single remote component per url, but I've not dug into the documentation.

Given that we'll soon be requiring a way to load and use multiple components from a remote bundle, I'm keen to have one solution here that works for the future rather than one solution here that works for a single remote component only, that we need to support while we add a separate one later (or have to break this support when updating).

Does that make sense?

Reply:
Ahhh okay I did not understand your concern at first but now I think I understand. We have all of our components compiled into 1 bundle, this includes all of the custom form components that we currently use in production today AND the custom app views that I've been developing while working on this feature. We use a logical case to determine what component we want to render. The entry point looks like:
```
import React from 'react';
import PropTypes from 'prop-types';
import DeploymentFormRenderer from './DeploymentForm/DeploymentFormRenderer';
import AppViewRenderer from './AppView/AppViewRenderer';

const componentIds = {
  AppViewRenderer,
  DeploymentFormRenderer,
};

const LocalComponent = props => {
  const Component = componentIds[props.componentType || 'DeploymentFormRenderer'];
  return <Component {...props} />;
};

LocalComponent.propTypes = {
  componentType: PropTypes.string,
};

export default LocalComponent;
```
And then in each renderer we determine what to render based on the props that it receives.

Keeping and rendering all components from a single bundle is possible and I think the cleanest/best way to do it. I want to add a `componentType` prop to the custom form components as well but I was going to do that in a separate PR to prevent scope creep.

Reply:
> Excellent - thanks Anthony, that's exactly what I was after. So the componentType prop is used in the developer provided code to return the component. So this means in effect that a class of views (AppView or DeploymentForm) can have a specific component. And as you say, the developer can do what they want here, so if you want a specific component for a specific chart, you can update the component selection to be dependent not only on the componentType but also the package identifier etc.

Yup! Exactly

> Which makes me then wonder, why do we need the customAppViews config at all... why not leave the decision wholly within the bundle? That is, what if you want to use your custom AppView for all packages? Currently the decision to use a custom app view is in both places: in the Kubeapps code depending on customAppViews, and also in the user provided bundle to decide which component to use. If a custom bundle had been provided, could we instead always check whether the bundle wants to provide the component and just default to the Kubeapps component when not? This way, the same code could be used in all parts of the UX (eg. in addition to AppView, we'd have RepositoryView, AppUpgrade, and for each, the bundle decides whether to return the default component or its own.

I decided to only have specific charts load the AppView for the time being since a developer might want to only inject these for certain deployments that require it. In my use case for example we leverage kubevirt to run VMs on our Kubernetes Cluster. The issue that we run into a lot is the delineation that they use for their resources (VirtualMachine and VirtualMachineInstance). We wanted to add information about the both resources. 

And I think there are some improvements that we can make for this sort of "plugin" architecture in the future to maybe make it simpler and the experience more streamlined. And I'm sure ill be able to come up with some cool ideas and places I will want to add some custom code :)